move **/tests/ to **/_tests/

jakkdl opened the pull request:
Simply renames the directories trio/tests and trio/_core/tests, and updates references to them.

Addresses some (or all?) problems in #274 
Fixes the second task in #2625 

@njsmith - especially wrt to considerations for #274

jakkdl commented:
ah, `trio/_tests/conftest.py` is not parsed for CLI flags (i.e. `--runslow`) before running. Could e.g. solve that by moving it into `trio/`, but that'd make it public.
There's probably other ways of achieving it ... but it's surprisingly tricky, and I've yet to manage something else decent

A5rocks commented:
I poked at this a bit and adding `-p "trio._tests.conftest"` seems to get further -- until it complains about adding the plugin twice (second time as normal part of test discovery).

Like the comment in the file says, we can essentially split `trio._tests.conftest` (which contains all of `trio._core._tests.conftest`) into a plugin and that'll work fine.

jakkdl commented:
> I poked at this a bit and adding `-p "trio._tests.conftest"` seems to get further -- until it complains about adding the plugin twice (second time as normal part of test discovery).
> 
> Like the comment in the file says, we can essentially put `trio._tests.conftest` (which contains all of `trio._core._tests.conftest`) into a plugin and that'll work fine.

huh, I don't get that far:
```python
  File "/home/h/Git/trio/trio/_tests/conftest.py", line 10, in <module>
    from ..testing import trio_test, MockClock
  File "/home/h/Git/trio/trio/testing/__init__.py", line 7, in <module>
    from ._sequencer import Sequencer
  File "/home/h/Git/trio/trio/testing/_sequencer.py", line 11, in <module>
    from .. import Event
ImportError: Error importing plugin "trio._tests.conftest": cannot import name 'Event' from 'trio' (unknown location)
```
though maybe I've messed up my environment somehow. Feel free to push a commit that makes it a plugin if you can get it to work.

jakkdl commented:
I'm still having some env issues with running the tests myself sometimes, but they seem to pass CI - so it should be fine on a clean install.

Added `trio/tests.py` and guessed how it was supposed to look :)

A5rocks reviewed (commented):
Presumably doing `import trio.tests` will work as a test but I don't think we need one.

jakkdl commented on trio/tests.py:
```diff
@@ -0,0 +1,10 @@
+from ._deprecate import warn_deprecated
+import trio
+from ._tests import *
```
I think the best way of handling that is by seeing that they're not listed in `pyright --verifytypes` in #2629 - and not do anything about it in this PR

jakkdl replied:
a test for `import trio.tests` would increase code coverage I suppose

jakkdl commented:
Cleaned up and added an import+deprecation test.
I'm a bit confused about codecov wrt `trio.testing`, but I don't think it's a problem?

Formatting seems to have hit a snag with ... something unrelated to this PR.
and 3.12 continues to fail like usual.

jakkdl commented:
@A5rocks this one is also ready for review

A5rocks reviewed (commented):
Cursory review comments, I might find something else in a second review!

A5rocks commented:
FYI master fixes the CI runs, so just merge that in!

jakkdl commented on trio/tests.py:
```diff
@@ -0,0 +1,36 @@
+    def __getattr__(self, attr: str) -> Any:
+        warn_deprecated(
+            f"trio.tests.{attr}",
+            "0.23.0",
```
Sounds good, I just copy-pasted another deprecation warning without considering the versioning :)

jakkdl commented:
rebased on top of master, and fixed review comments

A5rocks commented:
Sounds good, we want to redo codecov a bit too since it doesn't work with like Windows and whatever ATM. So feel free to change things slightly if necessary!

jakkdl commented:
uh, looks like codecov isn't turning up as a check at all now. I'll just revert it and we can put a pin into running with `codecov -m` for the upcoming codecov changes in a separate PR. I'd like this merged so it stops blocking #2629

A5rocks commented on trio/tests.py:
```diff
@@ -0,0 +1,36 @@
+import sys
+from typing import Any
+
+# don't ask which of the _tests imports is necessary
+import trio._tests
+
+from . import _tests
```
I'm a bit confused by this, won't the namespaces shadow and so the first is useless?

A5rocks commented on trio/tests.py:
```diff
@@ -0,0 +1,36 @@
+
+from . import _tests
+from ._deprecate import warn_deprecated
+from ._tests import *
```
... I don't think this is necessary anymore?

A5rocks reviewed (commented):
Looks mostly good, I'd like @njsmith to weigh in cause I'm not too used to changes like this!

jakkdl commented:
Fixed review comments, removing unnecessary imports. And reading the PR after having forgotten what the code does unsurprisingly lead to some improved comments and tests :innocent:

njsmith commented on trio/tests.py:
```diff
@@ -0,0 +1,31 @@
+
+
+# https://stackoverflow.com/questions/2447353/getattr-on-a-module
+sys.modules[__name__] = TestsDeprecationWrapper()  # type: ignore[assignment]
```
I'm not sure we really need to be this careful about backcompat here -- *surely* people aren't *actually* using `trio.tests`, right? -- but I guess it's harmless.

Though... I guess we actually want to get rid of `trio.tests` entirely, so maybe we should issue the warning on import? If someone is importing it but not using it (for some reason), then they won't get a warning, but they'll still get broken in the future when we delete this file.

jakkdl replied:
ah, took me a minute but figured out how to get it to warn on import as well. Only thing it doesn't warn on now that would break is `import trio; assert trio.tests` - i.e. access of the module, but without importing it, and not accessing any of it's attributes. (I tried changing from using `__getattr__` to `__getattribute__` but that made no difference in this case). There's probably some way of addressing that, but now we're getting into very weird cases.

jakkdl commented:
removed the test import from `__init__.py`, so it's now also invisible to static analysis. Added a corresponding exception for it being a missing symbols in the static analysis tests.

jakkdl commented:
@A5rocks any reservations about merging this?
